package/gluon-mesh-vpn-wireguard/luasrc/lib/gluon/upgrade/400-mesh-vpn-wireguard

```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/lua
+
+local uci = require('simple-uci').cursor()
+local util = require 'gluon.util'
+local site = require 'gluon.site'
+local hash = require 'hash'
+
+local function calc_address (public_key)
+  local s = hash.md5(public_key)
+  return 'fe80::02'..s:sub(1,2)..':'..s:sub(3,4)..'ff:fe'..s:sub(5,6)..':'..s:sub(7,10)
+end
+
+local private_key = uci:get("network", "vpn", "private_key")
+
+if not private_key or not private_key:match("^" .. ("[%a%d+/]"):rep(43) .. "=$") then
+  -- generate one
+  private_key = util.trim(util.exec("/usr/bin/wg genkey"))
```

Comment:
This is a bad idea, since the upgrade scripts seem to run before haveged is started. Not sure, whether this is intentional, but fastd seems to handle this by storing "generate" in the uci setting.

Leaving it like this is no option as this ocasionally bricks devices.

Reply:
Seems like the `wireguard` proto does not have an option like generate.

I don't know in which step we could do the generation otherwise. Ideas welcome.

Reply:
I submitted this patch upstream: https://patchwork.ozlabs.org/project/openwrt/patch/20210108012435.175345-1-me@irrelefant.net/